libvast/src/data.cpp

```diff
@@ -285,4 +338,108 @@ bool convert(const data& d, json& j, const type& t) {
   return convert(d, j);
 }
 
+namespace {
+
+data parse(const YAML::Node& node) {
+  switch (node.Type()) {
+    case YAML::NodeType::Undefined:
+    case YAML::NodeType::Null:
+      return data{};
+    case YAML::NodeType::Scalar: {
+      auto str = node.as<std::string>();
+      data result;
+      // Attempt some type inference.
+      if (parsers::boolean(str, result))
+        return result;
+      // Attempt maximum type inference.
+      if (parsers::data(str, result))
+        return result;
+      // Take the input as-is if nothing worked.
+      return str;
+    }
+    case YAML::NodeType::Sequence: {
+      list xs;
+      xs.reserve(node.size());
+      for (size_t i = 0; i < node.size(); ++i)
+        xs.push_back(parse(node[i]));
+      return xs;
+    }
+    case YAML::NodeType::Map: {
+      record xs;
+      xs.reserve(node.size());
+      for (auto& pair : node)
+        xs.emplace(pair.first.as<std::string>(), parse(pair.second));
+      return xs;
+    }
+  }
+  return {}; // makes GCC happy
+}
+
+} // namespace
+
+caf::expected<data> from_yaml(std::string_view str) {
+  try {
+    // Maybe one glory day in the future it will be possible to perform a
+    // single pass over the input without incurring a copy.
+    auto node = YAML::Load(std::string{str});
+    return parse(node);
```

Comment:
```suggestion
caf::expected<data> from_yaml(std::string str) {
  try {
    // Maybe one glory day in the future it will be possible to perform a
    // single pass over the input without incurring a copy.
    auto node = YAML::Load(std::move(str));
    return parse(std::move(node));
```

If a function must copy a parameter, it should take that parameter by value. Not a reference, not a view, but a value.

Imagine you have a `std::string` on the calling side that you don't need anymore after calling `from_yaml`. With the function taking a `std::string_view` and then copying it explicitly, you cannot avoid a copy. Furthermore, taking a `std::string_view` signals to the reader that the function does not copy its argument, which is misleading in this case.

Reply:
> If a function must copy a parameter, it should take that parameter by value. Not a reference, not a view, but a value.

The function signature in yaml-cpp looks like this:

```cpp
YAML_CPP_API Node Load(const std::string& input);
```

So moving in there doesn't strike me as an improvement.

My bigger point was though: there could be 10 copies happening internally, it doesn't matter because it's not going to be measurable. I don't want to change the API when an internal yaml-cpp optimization gets rolled out in the future. Therefore, `std::string_view` is still providing us with most flexibility at the call site.

Reply:
> Therefore, `std::string_view` is still providing us with most flexibility at the call site.

There is no flexibility here at all. The function is taking a string view only to create a new string from it. Let's make this clearer on our end regardless of the yaml-cpp API: Take a string when you want a string, take a string view when you want a string view. Here you want a string, not a string view. If I read this code a year down the road I might be confused as to why a copy is happening in a function that takes a string view.

Reply:
It looks like I'm not getting my point across very well, sorry for that. Let's ignore the implementation for a moment (because it doesn't matter performance-wise). In your opinion, what string type should a `to_yaml` function take as argument for maximum flexibility and modern practices?

Reply:
It should take a `std::string` if its implementation is required to copy and/or modify its input.

It should take a `std::string_view` if no copying occurs, ever.

Reply:
I think this is where we have a different approach: I wouldn't sacrifice API architecture because of an implementation detail *unless* it's on the critical path.

Reply:
Let's have @lava be the decider.

Reply:
I'd say let's not be too dogmatic for our own good:

```
caf::expected<data> from_yaml(const std::string& str)
```

* Never does a copy
* Makes sense from an API perspective
* Is actually the type that we want in the implementation

Reply:
>  if I have to choose between the two options, I'd go with keeping `string_view`

Okay, then I won't make any further API changes.